### ISSUE TITLE
Linking with mold fails

### ISSUE THREAD

**Issue by simeonmiteff** [zeek/spicy#1381]
I am experimenting with ways to speed up spicy builds, so I tried [mold](https://github.com/rui314/mold) which is supposedly the fastest linker in town.

With `spicyz` I get:

```
mold: fatal: cannot open dynamic_lookup: No such file or directory
```

I note that we're passing `-Wl,dynamic_lookup` to the linker: https://github.com/zeek/spicy/blob/da2696f904d1d7f7f902abb789860150073e45a6/hilti/toolchain/src/config.cc.in#L175

I presume this is necessary. Any idea if `mold` is broken?

**Comment by bbannier** [comment]
We use the `dynamic_lookup` flag to make sure that for precompiled HLTO files symbols are resolved when the HLTO file (which under the covers currently is just a shared library) is loaded, not when it is built. Looking over the mold source code I cannot see `dynamic_lookup` anywhere, so I suspect it does not support it yet (there is however a [related open issue](https://github.com/rui314/mold/issues/647)). It seems mold does not yet have enough support to be used to link Spicy modules.

> I am experimenting with ways to speed up spicy builds, so I tried [mold](https://github.com/rui314/mold) which is supposedly the fastest linker in town.

Alternatively, you could try the gold linker which we support AFAIK. That being said, for larger analyzers we typically spent more time JIT'ing source code or compiling generated code so I suspect it might not bring a huge speedup. You can check where time is spent by passing the `-R` flag to `spicyc`/`spicy-driver`/....

**Comment by simeonmiteff** [comment]
@bbannier thanks for looking into this!

Re: the linker's share of build time - I had a 16 hour (😲) build of a _giant_ analyser fail with GNU ld getting OOM-killed right at the end. I'm not sure if it is an ld problem specifically but I figured it can't help to pick a better linker.

@sethhall mentioned out-of-band that `lld` (Clang linker) is _also_ faster. I tested and can confirm that it works with spicy.

For the random traveller passing by, here is what I did on my Debian-flavoured Linux dev machine:

```shell
$ sudo apt-get install lld
$ sudo update-alternatives --install "/usr/bin/ld" "ld" "/usr/bin/ld.lld" 40
$ $ ld -v
Ubuntu LLD 14.0.0 (compatible with GNU linkers)
```
